objection_engine/utils.py

```diff
@@ -39,19 +44,53 @@ def get_characters(common: Counter):
         Character.OLDBAG,
         Character.REDD,
     ]
-    rnd_characters = []
-    if len(most_common) > 0:
-        users_to_characters[most_common[0]] = Character.PHOENIX
-        if len(most_common) > 1:
+
+    # If Phoenix was manually assigned, then assign him to his user.
+    # Otherwise, assign him to the user with the most comments.
+    if Character.PHOENIX in assigned_characters:
+        users_to_characters[assigned_characters[Character.PHOENIX]] = Character.PHOENIX
+        del assigned_characters[Character.PHOENIX]
+    else:
+        try:
+            users_to_characters[most_common[0]] = Character.PHOENIX
+        except IndexError:
+            pass
+
+    # Same for Edgeworth, but in the case of no manual assignment,
+    # assign him to the user with the second-most comments.
+    if Character.EDGEWORTH in assigned_characters:
+        users_to_characters[assigned_characters[Character.EDGEWORTH]] = Character.EDGEWORTH
+        del assigned_characters[Character.EDGEWORTH]
+    else:
+        try:
             users_to_characters[most_common[1]] = Character.EDGEWORTH
-            for character in most_common[2:]:
-                if len(rnd_characters) == 0:
-                    rnd_characters = all_rnd_characters.copy()
-                rnd_character = random.choice(
-                    rnd_characters
-                )
-                rnd_characters.remove(rnd_character)
-                users_to_characters[character] = rnd_character
+        except IndexError:
+            pass
+
+    # Before we start randomly assigning characters, let's match up all of the
+    # manually assigned characters to their user IDs.
+    for character, user_id in assigned_characters.items():
+        users_to_characters[user_id] = character
+        all_rnd_characters.remove(character)
```

Comment:
You shouldn't remove from `all_rnd_characters,`. `all_rnd_characters` is meant to include all characters, even used ones. This is because if there are more users in the conversation than total characters we get to recycle some characters.

So you should populate `rnd_characters` earlier and remove from there.

Unless you thought that once a user used a character nobody else should be able to use it which is a good point of view. What do you think it's better?

Reply:
My thought process here was that if a user explicitly assigned a character to a user, they wouldn't want that character being re-used for anyone else. That being said, this approach is inconsistent with how the rest of the characters are handled, so I think your approach would be better.